nomad/fsm_test.go

```diff
@@ -2836,6 +2848,183 @@ func TestFSM_ReconcileSummaries(t *testing.T) {
 	}
 }
 
+func TestFSM_ReconcileVersionedSummary(t *testing.T) {
+	logger := testlog.HCLogger(t)
+
+	ci.Parallel(t)
+	// Add some state
+	fsm := testFSM(t)
+	state := fsm.State()
+
+	// Add a node
+	node := mock.Node()
+	require.NoError(t, state.UpsertNode(structs.MsgTypeTestSetup, 800, node))
+
+	// Make a job
+	job1 := mock.Job()
+	job1.TaskGroups[0].Tasks[0].Resources.CPU = 5000
+
+	// Make a failed alloc on that job
+	alloc1 := mock.Alloc()
+	alloc1.NodeID = node.ID
+	alloc1.Job = job1
+	alloc1.JobID = job1.ID
+	alloc1.ClientStatus = structs.AllocClientStatusFailed
+
+	req := structs.GenericRequest{}
+	buf, err := structs.Encode(structs.ReconcileJobSummariesRequestType, req)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	require.NoError(t, state.UpsertJob(structs.MsgTypeTestSetup, 1000, nil, job1))
+	require.NoError(t, state.UpsertAllocs(structs.MsgTypeTestSetup, 1011, []*structs.Allocation{alloc1}))
+
+	require.NoError(t, state.DeleteJobSummary(1015, job1.Namespace, job1.ID))
+
+	resp := fsm.Apply(makeLog(buf))
+	if resp != nil {
+		t.Fatalf("resp: %v", resp)
+	}
+
+	ws := memdb.NewWatchSet()
+	out1, err := state.JobSummaryByID(ws, job1.Namespace, job1.ID)
+	require.NoError(t, err)
+
+	expected := structs.JobSummary{
+		JobID:     job1.ID,
+		Namespace: job1.Namespace,
+		Summary: map[string]structs.TaskGroupSummary{
+			"web": {
+				Queued: 10,
+				Failed: 1,
+			},
+		},
+		VersionedSummary: map[uint64]structs.VersionedSummary{
+			0: {
+				Version: 0,
+				Groups: map[string]structs.TaskGroupSummary{
+					"web": {
+						Failed: 1,
+					},
+				},
+			},
+		},
+		CreateIndex: 1000,
+		ModifyIndex: out1.ModifyIndex,
+	}
+	if !reflect.DeepEqual(&expected, out1) {
+		t.Fatalf("expected: %#v, actual: %#v", &expected, out1)
+	}
+
+	// Bump job to 1 and add a couple allocs
+
+	// alloc.ClientStatus = structs.AllocClientStatusRunning
+
+	// Create a new alloc here, have it be running
+
+	job2 := job1.Copy()
+	alloc2 := mock.Alloc()
+	alloc2.DesiredStatus = structs.AllocClientStatusRunning
+	alloc2.NodeID = node.ID
+	alloc2.Job = job2
+	alloc2.JobID = job2.ID
+	alloc2.ClientStatus = structs.AllocClientStatusRunning
+
+	// Create a new alloc here, have it be running
+	alloc3 := mock.Alloc()
+	alloc3.NodeID = node.ID
+	alloc3.Job = job2
+	alloc3.JobID = job2.ID
+	alloc3.ClientStatus = structs.AllocClientStatusRunning
+	require.NoError(t, state.UpsertJob(structs.MsgTypeTestSetup, 1020, nil, job2))
```

Comment:
We're gradually moving Nomad off of `testify` and towards `shoenig/test`, so let's use `must.NoError` for these kinds of assertions.